pkg/apis/serving/k8s_validation.go

```diff
@@ -82,8 +82,9 @@ var (
 )
 
 // ValidateVolumes validates the Volumes of a PodSpec.
-func ValidateVolumes(ctx context.Context, vs []corev1.Volume, mountedVolumes sets.String) (sets.String, *apis.FieldError) {
+func ValidateVolumes(ctx context.Context, vs []corev1.Volume, mountedVolumes sets.String) (sets.String, sets.String, *apis.FieldError) {
 	volumes := make(sets.String, len(vs))
+	nonReadOnlyVolumes := make(sets.String)
```

Comment:
thinking out loud: a nice approach might be to make the return value of this method a `map[string]corev1.Volume` (rather than two sets.Strings each with different semantics, which is maybe a bit unwieldy). The `volume.EmptyDir != nil` in this method would then not be needed at all, and in `validateVolumeMounts` you could do: `if volumes[vm.Name].EmptyDir != nil && !vm.ReadOnly` (which means the logic would only be in only one place).

Reply:
Yeah I also thought of returning volumes instead but didnt want to change the signature for existing code (less changes).

Reply:
Right but we have to change the signature anyway since we pass two sets.Strings around everywhere now?

Reply:
Yes but didnt touch some other parts though eg. access the first string set. Anyway I can take a look :)

Reply:
Heheheh yeah, worth a go anyway I think since it'd move all the logic to one place and avoid having two pass two things up and down the stack